pycbc/types/array_cpu.py

```diff
@@ -93,7 +102,62 @@ def vdot(self, other):
 def squared_norm(self):
     """ Return the elementwise squared norm of the array """
     return (self.data.real**2 + self.data.imag**2)
-    
+
+_ALLOWED_DTYPES = [_np.float32, _np.float64, _np.complex64,
+                   _np.complex128, _np.uint32]
+_ALLOWED_SCALARS = [int, long, float, complex] + _ALLOWED_DTYPES
+
+def multiply_and_add(self, mult_fac, add_fac, output):
+    """
+    Return input multiplied by mult_fac and with add_fac added.
+    If output is given this will also be the return value.
+    If not given array will be created internally.
+    This should be able to take real or complex mult_fac and add_fac,
+    potentially there will be issues if types are mixed, Ian can fix that
+    if errors are noticed.
+    """
+    # Sanity checking should have already be done. But we don't know if
+    # mult_fac and add_fac are arrays or scalars.
+    inpt = _np.array(self.data, copy=False)
+    N = len(inpt)
+    output = _np.array(output.data, copy=False)
+
+    if type(mult_fac) in _ALLOWED_SCALARS:
+        # For reasons I don't understand the code fails if this is a scalar and
+        # complex. Therefore force to an array. Precisions must match!!
+        # mult_index_str = ''
+        mult_fac = _np.array([mult_fac], dtype=type(mult_fac))
+        mult_index_str = '[0]'
+    else:
+        mult_index_str = '[i]'
+        mult_fac = _np.array(mult_fac.data, copy=False)
+
+    if type(add_fac) in _ALLOWED_SCALARS:
+        #add_index_str = ''
+        add_fac = _np.array([add_fac], dtype=type(add_fac))
+        add_index_str = '[0]'
+    else:
+        add_index_str = '[i]'
+        add_fac = _np.array(add_fac.data, copy=False)
+
+    code = """
+        #pragma omp parallel for
+        for (int i=0; i<N; i++){
+            output[i] = inpt[i] * mult_facMULTINDEX + add_facADDINDEX;
+        }
+    """
+    code = code.replace('MULTINDEX', mult_index_str)
+    code = code.replace('ADDINDEX', add_index_str)
```

Comment:
This string operation might be slow, especially for small Arrays. Any reason not to precompute these strings and look it up?